app.go

```diff
@@ -223,14 +229,15 @@ func (a *App) RunAsSubcommand(ctx *Context) (err error) {
 	// append help to commands
 	if len(a.Commands) > 0 {
 		if a.Command(helpCommand.Name) == nil && !a.HideHelp {
-			a.Commands = append(a.Commands, helpCommand)
-			if (HelpFlag != BoolFlag{}) {
+			a.appendCommand(helpCommand)
+
+			if HelpFlag != (&BoolFlag{}) {
```

Comment:
I think you want `*HelpFlag != BoolFlag{}` here lest it compare pointer addresses.